File: Applied Course/4.Problem Solving/11.Problem on Dynamic Programming/2.Longest Palindromic Subsequence.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
public:
    int longestPalindromeSubseq(string s) {
        int n=s.length();
        int LPS[n+1][n+1];
        for(int i=0;i<n;i++)
		    LPS[i][i]=1;
        for(int ls=2;ls<=n;ls++)
        {
            for(int i=0;i<=n-ls;i++)
            {
                int j=i+ls-1;
                if(s[i]==s[j] && ls==2)
                    LPS[i][j]=2;
                else if(s[i]==s[j])
                    LPS[i][j]=LPS[i+1][j-1]+2;
                else
                    LPS[i][j]=max(LPS[i][j-1],LPS[i+1][j]);
            }
        }
        return LPS[0][n-1];
    }
};
int main()
{
    // https://leetcode.com/problems/longest-palindromic-subsequence/
    return 0;
}

File: Applied Course/4.Problem Solving/11.Problem on Dynamic Programming/2.Longest Palindromic sub sequence.py
# https://leetcode.com/problems/longest-palindromic-subsequence/
# Logic-1 DP
class Solution:
    def longestPalindromeSubseq(self, s: str) -> int:
        n = len(s) 
        dp = [[0 for x in range(n)] for x in range(n)] 
        for i in range(n): 
            dp[i][i] = 1    
        for x in range(2, n+1): 
            for i in range(n-x+1): 
                j = i+x-1
                if s[i] == s[j] and x == 2: 
                    dp[i][j] = 2
                elif s[i] == s[j]: 
                    dp[i][j] = dp[i+1][j-1] + 2
                else: 
                    dp[i][j] = max(dp[i][j-1], dp[i+1][j])
        return dp[0][n-1] 
# Logic-2 Recursion
def longestPalindromicSubsequnce(s,i,j):
    if i==j:
        return 1
    if s[i]==s[j] and i+1==j:
        return 2
    if s[i]==s[j]:
        return longestPalindromicSubsequnce(s,i+1,j-1)
    
    return max(longestPalindromicSubsequnce(s,i+1,j),longestPalindromicSubsequnce(s,i,j-1))